R/bookmark-state.R

```diff
@@ -0,0 +1,960 @@
+#' copied. The result from \code{urlModal} should be passed to the
+#' \code{\link{showModal}} function to display it in the browser.
+#'
+#' @param url A URL to display in the dialog box.
+#' @param title A title for the dialog box.
+#' @param subtitle Text to display underneath URL.
+#' @export
+urlModal <- function(url, title = "Bookmarked application link", subtitle = NULL) {
+
+  subtitleTag <- NULL
+  if (!is.null(subtitle)) {
+    subtitleTag <- tagList(
+      br(),
+      span(class = "text-muted", subtitle),
+      span(id = "shiny-bookmark-copy-text", class = "text-muted")
+    )
+  }
+
+  modalDialog(
+    title = title,
+    easyClose = TRUE,
+    footer = NULL,
+    tags$textarea(class = "form-control", rows = "1", style = "resize: none;",
+      readonly = "readonly",
+      url
+    ),
+    subtitleTag,
+    # Need separate show and shown listeners. The show listener sizes the
+    # textarea just as the modal starts to fade in. The 200ms delay is needed
+    # because if we try to resize earlier, it can't calculate the text height
+    # (scrollHeight will be reported as zero). The shown listener selects the
+    # text; it's needed because because selection has to be done after the fade-
+    # in is completed.
+    tags$script(
+      "$('#shiny-modal').
+        one('show.bs.modal', function() {
+          setTimeout(function() {
+            var $textarea = $('#shiny-modal textarea');
+            $textarea.innerHeight($textarea[0].scrollHeight);
+          }, 200);
+        });
+      $('#shiny-modal')
+        .one('shown.bs.modal', function() {
+          $('#shiny-modal textarea').select().focus();
+        });
+      $('#shiny-bookmark-copy-text')
+        .text(function() {
+          if (/Mac/i.test(navigator.userAgent)) {
+            return 'Press \u2318-C to copy.';
+          } else {
+            return 'Press Ctrl-C to copy.';
+          }
+        });
+      "
+    )
+  )
+}
+
+#' Enable bookmarking for a Shiny application
+#'
+#' There are two types of bookmarking: saving an application's state to disk on
+#' the server, and encoding the application's state in a URL. For state that has
+#' been saved to disk, the state can be restored with the corresponding state
+#' ID. For URL-encoded state, the state of the application is encoded in the
+#' URL, and no server-side storage is needed.
+#'
+#' For restoring state to work properly, the UI must be a function that takes
+#' one argument, \code{request}. In most Shiny applications, the UI is not a
+#' function; it might have the form \code{fluidPage(....)}. Converting it to a
+#' function is as simple as wrapping it in a function, as in
+#' \code{function(request) \{ fluidPage(....) \}}.
+#'
+#' By default, all input values will be bookmarked, except for the values of
+#' actionButtons and passwordInputs. FileInputs will be saved if the state is
+#' saved on a server, but not if if the state is encoded in a URL.
```

Comment:
Double "if". Also I wouldn't capitalize "fileInput". And these three functions should all be \code{\link{..}}